File: test17_out/pollen.time/TimerManager/TimerManager.js

/*
 * ======== MODULE pollen.time.TimerManager ========
 */

var $$c = function() {
    const TimerManager = {};
    TimerManager.TimerManager = TimerManager
    TimerManager.$name = 'pollen.time.TimerManager';
    TimerManager.pollen$used = 0;
    TimerManager.Event = $units['pollen.events.Event'];
    TimerManager.Timer$$id = 1;
    TimerManager.Timer = function(cn) {
        this.$$id = TimerManager.Timer$$id++;
        this.$$cname = cn ? cn : 'pollen_time_TimerManager_Timer__'+this.$$id+'__S';
        this.$$tname = 'pollen_time_TimerManager_Timer';
        this.$$uname = 'pollen.time.TimerManager';
        this.id = 0;
        this.duration = 0;
        this.elapsed = 0;
        this.repeat = 0;
    }
    TimerManager.Timer.prototype = new $$Struct('pollen.time.TimerManager.Timer');
    TimerManager.Timer.prototype.$$isAggFld = {};
    TimerManager.Timer.prototype.$$isAggFld.id = false;
    TimerManager.Timer.prototype.$$isAggFld.expired = false;
    TimerManager.Timer.prototype.$$isAggFld.duration = false;
    TimerManager.Timer.prototype.$$isAggFld.elapsed = false;
    TimerManager.Timer.prototype.$$isAggFld.repeat = false;
    TimerManager.Timer.prototype.init = function() {
        var $$text = '';
        (this.id) = (TimerManager.numTimers);
        (TimerManager.numTimers)++;
        (this.repeat) = false;
        $$printf("\nTIMER INIT");
        this.$$hostInitLastPass();
        return $$text;
    }
    TimerManager.Timer.prototype.new_host = function() {
        var $$text = '';
        this.$$privateInit();
        this.expired = 0;
        this.$$hostInitLastPass();
        return this;
    }
    TimerManager.Timer.prototype.$$privateInit = function() {
        this.id = 0;
        this.duration = 0;
        this.elapsed = 0;
        this.repeat = 0;
    }
    TimerManager.Timer.prototype.$$hostInitLastPass = function() {
    }
    TimerManager.timerISR = new $$Ref('pollen_time_TimerManager_timerISR__I');
    TimerManager.targetInit = new $$Ref('pollen_time_TimerManager_targetInit__I');
    TimerManager.bindGlobalInterrupts = function( gi ) {
        var $$text = '';
        var $$v = 'pollen_interfaces_GlobalInterrupts';
        var $$s = $$bind($units['pollen.time.TimerManager'], 'gint', $$v);if ($$v && $$v == $$s) $$v.pollen$used = true;

        return $$text;
    }
    TimerManager.bindTimerDriver = function( t ) {
        var $$text = '';
        var $$v = 'pollen_interfaces_TimerDriver';
        var $$s = $$bind($units['pollen.time.TimerManager'], 'hw_timer', $$v);if ($$v && $$v == $$s) $$v.pollen$used = true;

        return $$text;
    }
    TimerManager.$$hostInit = function() {
        var $$text = '';
        return $$text;
    }
    TimerManager.pollen__uses__ = function() {
        $$bind($units['pollen.events.Dispatcher'], 'pollen$used', true);
        $$bind($units['pollen.events.Event'], 'pollen$used', true);
    }
    TimerManager.$$privateInit = function() {
        TimerManager.gint = 0;
        TimerManager.hw_timer = 0;
        TimerManager.numTimers = 0;
    }
    TimerManager.$$hostInitLastPass = function() {
        TimerManager.timers = new $$Array((TimerManager.numTimers), function($$cn,$$idx){obj = new TimerManager.Timer(); obj.new_host(); return obj;}, 'pollen_time_TimerManager_timers__A', false);
    }
    return TimerManager;
}
var $$u = $$c();
$units['pollen.time.TimerManager'] = $$u;
$units.push($$u);

File: test17_out/pollen.time/TimerManager/TimerManager.c

/*
 * ======== string literals ========
 */

static const char pollen_time_TimerManager__s__fileName[] = "TimerManager.p";
static const char pollen_time_TimerManager__s__0[] = "\nTIMER INIT";
static const char pollen_time_TimerManager__s__1[] = "Timer";
static const char pollen_time_TimerManager__s__2[] = "TimerManager";

/*
 * ======== function definitions ========
 */

bool pollen_time_TimerManager_Timer_repeats__E( pollen_time_TimerManager_Timer_ this ) {
    return(  (this->repeat)  );
}

void pollen_time_TimerManager_Timer_start__E( pollen_time_TimerManager_Timer_ this, struct pollen_events_Event*  e, uint16 d, bool r ) {
    bool state;
    state = pollen_interfaces_GlobalInterrupts_disable__E();
    if (!pollen_time_TimerManager_Timer_running__E(this)) {
        (this->expired) = (e);
        (this->duration) = (d);
        (this->elapsed) = 0;
        (this->repeat) = (r);
        (pollen_time_TimerManager_timers__A[(this->id)]) = this;
    }
    pollen_interfaces_GlobalInterrupts_restore__E((state));
    if (!pollen_interfaces_TimerDriver_running__E()) {
        pollen_interfaces_TimerDriver_start__E();
    }
}

pollen_time_TimerManager_Timer_ pollen_time_TimerManager_Timer_new___E( pollen_time_TimerManager_Timer_ this ) {
    return this;
}

bool pollen_time_TimerManager_Timer_running__E( pollen_time_TimerManager_Timer_ this ) {
    return(  (pollen_time_TimerManager_timers__A[(this->id)]) != null  );
}

void pollen_time_TimerManager_Timer_setElapsed__E( pollen_time_TimerManager_Timer_ this, uint32 i ) {
    (this->elapsed) = (i);
}

void pollen_time_TimerManager_Timer_stop__E( pollen_time_TimerManager_Timer_ this ) {
    bool state;
    state = pollen_interfaces_GlobalInterrupts_disable__E();
    (pollen_time_TimerManager_timers__A[(this->id)]) = null;
    pollen_interfaces_GlobalInterrupts_restore__E((state));
}

void pollen_time_TimerManager_timerISR__I() {
    bool off;
    uint8 i;
    off = true;
    for (i = 0; (i) < (pollen_time_TimerManager_numTimers__V); (i)++) {
        if ((pollen_time_TimerManager_timers__A[(i)]) != null) {
            (pollen_time_TimerManager_timers__A[(i)]->elapsed)++;
            if ((pollen_time_TimerManager_timers__A[(i)]->elapsed) == (pollen_time_TimerManager_timers__A[(i)]->duration)) {
                if ((pollen_time_TimerManager_timers__A[(i)]->expired) != null) {
                    pollen_events_Dispatcher_postFromInterrupt__E((pollen_time_TimerManager_timers__A[(i)]->expired));
                    if ((pollen_time_TimerManager_timers__A[(i)]->repeat)) {
                        (pollen_time_TimerManager_timers__A[(i)]->elapsed) = 0;
                    }
                    else {
                        pollen_time_TimerManager_Timer_stop__E((pollen_time_TimerManager_timers__A[(i)]));
                    }
                }
                else {
                    pollen_time_TimerManager_Timer_stop__E((pollen_time_TimerManager_timers__A[(i)]));
                }
            }
            (off) = false;
        }
    }
    if ((off)) {
        pollen_interfaces_TimerDriver_stop__E();
    }
}

void pollen_time_TimerManager_targetInit__I() {
}

uint16 pollen_time_TimerManager_Timer_ticksPerSecond__E( pollen_time_TimerManager_Timer_ this ) {
    return(  1000  );
}

uint32 pollen_time_TimerManager_Timer_getElapsed__E( pollen_time_TimerManager_Timer_ this ) {
    return(  (this->elapsed)  );
}

